### DISCUSSION TITLE
How does element suggestion work?

### DISCUSSION THREAD

**Discussion by 5e-Cleric** [webextensions/live-css-editor#141] [Q&A] [answered]
Hi, i am working on a web app and using CodeMirror 6 to style stuff inside it. 

I was wondering how do you manage to provide a list of selectors to suggest to the user, do you use an extension? how do you create the dynamic list of selectors?

**Comment by webextensions** [comment, marked as answer]
@5e-Cleric 

It is custom code. There are 2 parts for suggesting the CSS selectors:

1. **Getting existing selectors:** Defined in function `getExistingCSSSelectors` at: https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/extension/scripts/magicss/magicss.js#L1078
2. **Generating selectors:** Defined in function `generateSelector` in file https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/extension/scripts/magicss/generate-selector.js which is used at: https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/extension/scripts/magicss/magicss.js#L2158

There is an intention to move these aspects into their own libraries when time permits, as noted at: https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/TODO.md?plain=1#L34

**Reply by 5e-Cleric** [reply]
If they were libraries that would be awesome, meanwhile thanks for all the info! Have you considered making the entire code esitor with the extra features as an entire library or package? i would consider it as an alternative to codeMirror. Specially with the less and scss compatibility.

**Reply by webextensions** [reply]
@5e-Cleric
Thank you for your appreciation!

We may move such reusable functions into npm packages of their own (or at least export them from some package).

While those functions are not well documented, we hope that those functions should be robust enough and can be copy-pasted by someone who wishes to utilize those.

From time-to-time, we move some of our reusable functions / wrappers under https://www.npmjs.com/package/helpmate and https://www.npmjs.com/package/helpmate-css from our different projects (though without any documentation so far, apart from a decent enough function name).

This editor extensively utilizes CodeMirror (v5) and its various plugins. We had to customize some aspects of those plugins to enable some of those auto-complete features (Ref: https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/extension/scripts/3rdparty/codemirror/addons/hint/css-hint_customized.js.source.txt and other such customized files near that location).

The floating editor in itself (with customizable toolbar etc) may be made as a reusable project, as intended at: https://github.com/webextensions/live-css-editor/blob/ad07e6fe50d1c7684eab8313fe760cf70ccdafd1/TODO.md?plain=1#L26
But, that would need a more dedicated effort, bigger refactoring and other practical decisions (eg: should that be rewritten in React?). Hence, not expecting that in near future.

**Reply by 5e-Cleric** [reply]
> While those functions are not well documented, we hope that those functions should be robust enough and can be copy-pasted by someone who wishes to utilize those.

I'll probably have to write them from scratch for my project, because the structure is different, and we don't use jQuery, but it is a great place to start.
